test/persistence-hooks.suite.js

```diff
@@ -1374,6 +1374,139 @@ module.exports = function(dataSource, should) {
         });
       });
     });
+    
+    describe('PersistedModel.prototype.replaceAttributes', function() {
+      it('triggers hooks in the correct order', function(done) {
+        monitorHookExecution();
+
+        existingInstance.replaceAttributes(
+          { name: 'replaced' },
+          function(err, record, created) {
+            if (err) return done(err);
+            triggered.should.eql([
+              'before save',
+              'persist',
+              'loaded',
+              'after save'
+            ]);
+            done();
+          });
+      });
+
+      it('triggers `before save` hook', function(done) {
+        TestModel.observe('before save', pushContextAndNext());
+
+        existingInstance.replaceAttributes({ name: 'changed' }, function(err) {
+          if (err) return done(err);
+          observedContexts.should.eql(aTestModelCtx({
+            instance: {
+              id: existingInstance.id,
+              name: 'changed',
+              extra: undefined
+            }
+          }));
+          done();
+        });
+      });
+
+      it('aborts when `before save` hook fails', function(done) {
+        TestModel.observe('before save', nextWithError(expectedError));
+
+        existingInstance.replaceAttributes({ name: 'replaced' }, function(err) {
+          [err].should.eql([expectedError]);
+          done();
+        });
+      });
+
+      it.skip('applies updates from `before save` hook', function(done) {
+        /* Should they be able to change `instance` in `before save`? I was completely on the wrong track? :(
+         * in line [1] data is over writting instance!
+         * Also in line [2] I was copying `data` to `inst`
+         * Could you please clarify if in `before save` for `replaceAttributes` they are allowed to change instance? 
+         * 
+         * [1]:https://github.com/strongloop/loopback-datasource-juggler/pull/788/files#diff-d717a01e40d8164976e4468af1bce663R2678
+         * [2]:https://github.com/strongloop/loopback-datasource-juggler/pull/788/files#diff-d717a01e40d8164976e4468af1bce663R2707
+        */
```

Comment:
Yes, it's the core purpose of before save hook to allow observers to modify instance data (properties, attributes). The test case looks legitimate to me and should be passing.

The links `[1]` and `[2]` don't work for me :(